### PR TITLE
Dont decrypt shared files

### LINKED ISSUE #25599: OCC Decrypt all always fails to complete.
Firstly let me say that Owncloud is brilliant..

When I first installed Owncloud back in the 8.x days I didn't properly grasp how the encryption module works and that locally encrypted files are a little pointless since the keys are sitting there too.. 100% of our storage is local so I really shouldn't have even enabled the module in the first place..

For quite a while now I have been trying to run the decrypt-all OCC command to remove encryption from the local files but it always fails.. I upgraded to OC 9.1 over the weekend and tried again but it still failed.. There are no specific error messages it simply appears to stop after a number of hours..

Another issue that seems to be causing the process to run a LOT longer than it needs to is the attempted re-decrypting of shared files over and over again for each user who has access to those files.. 99% of the files stored on the server are shared between most/all of the users and the decryption is running through the same files over and over again.. Surely if the file has been decrypted in step 1 of x it doesn't need to be re-decrypted for each user that has access to that file? 

Perhaps I need to decrypt user by user individually then i won't try re-decrypt the files? Issue there is that it seems the first step is to disable the encryption module so not sure how that would work out.. Or maybe I have to un-share all files then decrypt-all and then re-share the files again but that too would be quite a job and would cause a lot of resyncing of data to the PC's..

The documentation doesn't suggest there is a special process to follow to remove and disable encryption..

### Steps to reproduce
1. run ./occ encryption:decrypt-all

### Expected behaviour
All files for all users decrypted and encryption module no longer required.

### Actual behaviour
Process starts and runs for a very long time through step 1 of 7, 2 of 7.... Then it falls over.. 
Last attempt made it to 6 of 7 before stopping but other attempts haven't made it as far.. This took about 3 hours.. All files then totally inaccessible through the web interface so I then have to completely restore the server from backup..

### Server configuration
**Operating system**: 
Cpanel server running Centos

**Web server:**
Apache under Cpanel

**Database:**
Mysql under Cpanel

**PHP version:**
5.6.20

**ownCloud version:** (see ownCloud admin page)
Currently 9.1 but have had the issue through all the 9.0.x versions since I started trying to remove the encryption module.

**Updated from an older ownCloud or fresh install:**
Upgraded

**Where did you install ownCloud from:**
Owncloud website

**Signing status (ownCloud 9.0 and above):**
No errors found

**List of activated apps:**
Enabled:
  - activity: 2.3.2
  - comments: 0.3.0
  - dav: 0.2.5
  - encryption: 1.3.0
  - federatedfilesharing: 0.3.0
  - files: 1.5.1
  - files_pdfviewer: 0.8.1
  - files_sharing: 0.10.0
  - files_texteditor: 2.1
  - files_trashbin: 0.9.0
  - files_versions: 1.3.0
  - files_videoplayer: 0.9.8
  - firstrunwizard: 1.1
  - notifications: 0.3.0
  - provisioning_api: 0.5.0
  - templateeditor: 0.1
  - updatenotification: 0.2.1
Disabled:
  - external
  - federation
  - files_antivirus
  - files_external
  - gallery
  - systemtags
  - user_external
  - user_ldap

**The content of config/config.php:**

Default config

**Are you using external storage, if yes which one:** local/smb/sftp/...
No

**Are you using encryption:** yes/no
Yes

**Are you using an external user-backend, if yes which one:** LDAP/ActiveDirectory/Webdav/...
No

### Client configuration
**Browser:**
SSH

### PR DESCRIPTION
**Pull request by DeepDiver1975**
fixes https://github.com/owncloud/core/issues/25599

THX @schiessle 

@wipeoutdude  mind testing? THX

### CODE REVIEW

**Comment by wipeoutdude** [comment]
@DeepDiver1975 - I will have a go at testing it at the next opportunity I have and feedback the results.

**Comment by VicDeo** [comment]
no way...
Am I doing something wrong?

1. Install OC from master
2. Enable default encryption module
3. Enable serverside encryption and set a recovery key
4. Create a user 'P'
5. Relogin
6. Login as P
7. Share welcome.txt with P
8. php  occ encr:encrypt-all
9. php occ encr:decrypt-all


```
Files for following users couldn't be decrypted, 
maybe the user is not set up in a way that supports this operation: 
    D
    P

deo@jah-mobile:~/public_html/oc-tmp> cat data/owncloud.log 

{"reqId":"PJGCahQhbUHGI4GyUTPs","remoteAddr":"","app":"PHP","message":"fclose() expects parameter 1 to be resource, string given at \/home\/deo\/public_html\/oc-tmp\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php#748","level":3,"time":"2016-08-08T13:25:32+00:00","method":"--","url":"--","user":"--"}
{"reqId":"PJGCahQhbUHGI4GyUTPs","remoteAddr":"","app":"PHP","message":"fclose() expects parameter 1 to be resource, string given at \/home\/deo\/public_html\/oc-tmp\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php#749","level":3,"time":"2016-08-08T13:25:32+00:00","method":"--","url":"--","user":"--"}
{"reqId":"PJGCahQhbUHGI4GyUTPs","remoteAddr":"","app":"PHP","message":"stream_wrapper_register(): Protocol ocencryption:\/\/ is already defined. at \/home\/deo\/public_html\/oc-tmp\/lib\/private\/Files\/Stream\/Encryption.php#195","level":3,"time":"2016-08-08T13:25:32+00:00","method":"--","url":"--","user":"--"}
{"reqId":"PJGCahQhbUHGI4GyUTPs","remoteAddr":"","app":"PHP","message":"fclose() expects parameter 1 to be resource, string given at \/home\/deo\/public_html\/oc-tmp\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php#748","level":3,"time":"2016-08-08T13:25:32+00:00","method":"--","url":"--","user":"--"}
{"reqId":"PJGCahQhbUHGI4GyUTPs","remoteAddr":"","app":"PHP","message":"fclose() expects parameter 1 to be resource, string given at \/home\/deo\/public_html\/oc-tmp\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php#749","level":3,"time":"2016-08-08T13:25:32+00:00","method":"--","url":"--","user":"--"}
```

File: lib/private/Encryption/DecryptAll.php [hunk 1]
```diff
@@ -209,6 +209,10 @@ protected function decryptUsersFiles($uid, ProgressBar $progress, $userCount) {
 		while ($root = array_pop($directories)) {
 			$content = $this->rootView->getDirectoryContent($root);
 			foreach ($content as $file) {
+				// only decrypt files owned by the user
+				if($file->getStorage()->instanceOfStorage('OC\Files\Storage\Shared')) {
```
Comment (PVince81) [inline lib/private/Encryption/DecryptAll.php]: now thinking of it, what would happen with global external storages ? @schiessle

**Comment by PVince81** [comment]
@VicDeo weird stuff you're getting. Maybe you found a bug ?

I simply used the user's login password for decryption with new files, no "encrypt-all" and this PR worked fine as it left the shared files alone :+1:

File: lib/private/Encryption/DecryptAll.php (hunk 1 above)
Comment (DeepDiver1975) [inline lib/private/Encryption/DecryptAll.php]: no idea - we better test this ....

**Comment by PVince81** [comment]
@VicDeo please retry

**Comment by VicDeo** [comment]
@PVince81 I already tried  over 6 times. Is anything changed here?

**Comment by PVince81** [comment]
@VicDeo nothing changed, I just hoped you mistyped your steps.
Does it work if you use the user's login password instead and skip "encrypt-all" but use a file that got encrypted through upload ?

I'll also try with your exact steps.

**Comment by PVince81** [comment]
@VicDeo can't reproduce the issue with your steps. As they are missing some slight details, here are mine with details:

1. Setup OC on this branch (from git without extra apps)
1. Enable default encryption module in apps page
1. Enable encryption in admin page
1. Create a user "test"
1. Log out, log in again as admin
1. Set a recovery password in admin page
1. Login as "test"
1. Go to personal page and opt-in with "enable recovery key" (this piece was missing in your steps)
1. Login as "admin"
1. Share "welcome.txt" (the non-encrypted skeleton file) with user "test"
1. Logout
1. On the CLI, run `./occ encryption:encrypt-all`, say "y" to continue
1. On the CLI, run `./occ encryption:decrypt-all`, say "y" to continue and enter recovery password

```
± % sudo -uwwwrun ./occ encryption:decrypt-all
The process control (PCNTL) extensions are required in case you want to interrupt long running commands - see http://php.net/manual/en/book.pcntl.php
Disable server side encryption... done.


You are about to start to decrypt all files stored in your ownCloud.
It will depend on the encryption module and your setup if this is possible.
Depending on the number and size of your files this can take some time
Please make sure that no user access his files during this process!

Do you really want to continue? (y/n) y
prepare encryption modules...

Prepare "Default encryption module"

You can only decrypt the files of all users if the
recovery key is enabled by the admin and activated by the users.

Please enter the recovery key password: 
 done.

 Fetch list of users... finished 
 [============================]

 starting to decrypt files... finished                           
 [============================]                                  


Files for following users couldn't be decrypted, 
maybe the user is not set up in a way that supports this operation: 
    admin
```

@VicDeo can you check whether your steps are exactly the same or have differences ?
Also, please run your steps directly on master to find out whether the issue you see is related to this PR or another issue.

Thanks.

**Comment by VicDeo** [comment]
@PVince81 confirmed. 
Both `admin` and `test` user should have `Enable password recovery` set to `enabled`

**Comment by PVince81** [comment]
@VicDeo ah... in my case I only enabled recovery for the test user, not admin

**Comment by VicDeo** [comment]
:+1: 
 maybe it worth to check the user settings..

```
Files for following users couldn't be decrypted, 
maybe the user is not set up in a way that supports this operation: 
    D
```
```
Files for following users couldn't be decrypted: 
    D - password recovery disabled
```
or something like this....

**Comment by PVince81** [comment]
Rebased for CI.

**Comment by PVince81** [comment]
Failing test unrelated. Raised as https://github.com/owncloud/core/issues/25830 as it fails randomly.

**Comment by PVince81** [comment]
stable9.1: https://github.com/owncloud/core/pull/25831
stable9: https://github.com/owncloud/core/pull/25832

**Pull request merged**